Tests/CacheAdvanceTests/CacheAdvanceTests.swift

```diff
@@ -96,6 +96,35 @@ final class CacheAdvanceTests: XCTestCase {
         XCTAssertEqual(messages, [])
     }
 
+  func test_messages_throwsFileCorruptedWhenOffsetInFileAtEndOfNewsetMessageOutOfSync() throws {
```

Comment:
> The only way this could happen is if the writer (and offset of the start of the oldest message) was at the end of the file's header

I agree with this analysis!

> If we were in this situation, `cacheHasSpaceForNewMessageBeforeEndOfFile` could never be `false`. If `cacheHasSpaceForNewMessageBeforeEndOfFile` were false we would have already failed this previous check, since in this scenario `writer.offsetInFile` would equal `FileHeader.expectedEndOfHeaderInFile`

I disagree!

If we were writing a message of length `header.maximumBytes - FileHeader.expectedEndOfHeaderInFile`, then  `bytesNeededToStoreMessage <= header.maximumBytes - FileHeader.expectedEndOfHeaderInFile` would evaluate as `true` since `bytesNeededToStoreMessage` is equal to `header.maximumBytes - FileHeader.expectedEndOfHeaderInFile`.

If `writer.offsetInFile` is at the `FileHeader.expectedEndOfHeaderInFile` (which I agree is a precondition for this bug), then in the next line `cacheHasSpaceForNewMessageBeforeEndOfFile` would evaluate to `true` because `writer.offsetInFile + bytesNeededToStoreMessage` equals `header.maximumBytes`:
https://github.com/dfed/CacheAdvance/blob/ed9ece4a186b0073514128a75f1cc401ba4ac130/Sources/CacheAdvance/CacheAdvance.swift#L113

At which point, I think we've proved that our original analysis was correct. Am I missing something?

To dig in a bit further, I believe this issue occurred because:
1. `offsetInFileAtEndOfNewestMessage` was beyond the actual end of the file
2. The file was empty except for the header

We only update `offsetInFileAtEndOfNewestMessage` after writing the message: https://github.com/dfed/CacheAdvance/blob/ed9ece4a186b0073514128a75f1cc401ba4ac130/Sources/CacheAdvance/CacheAdvance.swift#L238-L246

Which means that the only way for `offsetInFileAtEndOfNewestMessage` to be beyond the actual end of the file is if we deleted part of the file before updating the header with our next write.

Now, this analysis (thank you for making me write this out!) indicates that my fix to prevent this data corruption in the future was incorrect. The issue isn't when we truncate the file, but rather when we update `offsetInFileAtEndOfNewestMessage`. PR incoming!

Reply:
https://github.com/dfed/CacheAdvance/pull/58

Reply:
This explanation makes sense. It seems like I was off by one. Thank you very much for writing this out.

Reply:
@dfed I made a correction to my original comment. Can you verify that my correction is in fact correct? I am trying to make sure I have the right mental model.

Reply:
> > If we were in this situation, `cacheHasSpaceForNewMessageBeforeEndOfFile` could never be `false`. If `cacheHasSpaceForNewMessageBeforeEndOfFile` were false we would have already failed this previous check, since in this scenario `writer.offsetInFile` would equal `FileHeader.expectedEndOfHeaderInFile`

> I disagree!

> If we were writing a message of length `header.maximumBytes - FileHeader.expectedEndOfHeaderInFile`, then  `bytesNeededToStoreMessage <= header.maximumBytes - FileHeader.expectedEndOfHeaderInFile` would evaluate as `true` since `bytesNeededToStoreMessage` is equal to `header.maximumBytes - FileHeader.expectedEndOfHeaderInFile`.

> If `writer.offsetInFile` is at the `FileHeader.expectedEndOfHeaderInFile` (which I agree is a precondition for this bug), then in the next line `cacheHasSpaceForNewMessageBeforeEndOfFile` would evaluate to `true` because `writer.offsetInFile + bytesNeededToStoreMessage` equals `header.maximumBytes`:
https://github.com/dfed/CacheAdvance/blob/ed9ece4a186b0073514128a75f1cc401ba4ac130/Sources/CacheAdvance/CacheAdvance.swift#L113

I thought more about this and I've circled back to my original conclusion. Yes, I agree that `cacheHasSpaceForNewMessageBeforeEndOfFile` would evaluate to `true`. However, we only truncate the file if `cacheHasSpaceForNewMessageBeforeEndOfFile` is `false`.

Reply:
We talked on Zoom and agreed with where I landed in my last comment.